File: web/src/services/api/finance.ts
import type { ListResponse } from "@/types/pagination";
import type { UUID } from "@/types/primitive";
import { http } from "./client";
import { ENDPOINTS } from "./endpoints";

export interface FinanceAsset {
  id: UUID;
  code: string;
  name?: string | null;
  decimal_places: number;
  is_default: boolean;
}

export interface FinanceTreeNode {
  id: UUID;
  parent_id: UUID | null;
  name: string;
  currency_code: string | null;
  path: string;
  depth: number;
  display_order: number;
}

export interface FinanceTree {
  id: UUID;
  name: string;
  primary_currency: string;
  display_order: number;
  is_default: boolean;
  nodes?: FinanceTreeNode[];
}

export interface FinanceSnapshotEntry {
  id: UUID;
  node_id: UUID;
  node_name: string | null;
  amount: string;
  currency_code: string;
  amount_converted: string;
  note?: string | null;
  is_auto_generated: boolean;
}

export interface FinanceSnapshot {
  id: UUID;
  tree_id: UUID;
  tree_name?: string | null;
  title?: string | null;
  snapshot_ts: string | null;
  period_start: string | null;
  period_end: string | null;
  primary_currency: string;
  rate_snapshot_id?: UUID | null;
  exchange_rates?: Record<string, unknown> | null;
  summary?: Record<string, unknown> | null;
  note?: string | null;
  entries?: FinanceSnapshotEntry[];
  created_at: string;
}

export interface FinanceRateSnapshotEntry {
  id: UUID;
  base_currency: string;
  quote_currency: string;
  rate: string;
  source?: string | null;
  captured_at?: string | null;
}

export interface FinanceRateSnapshot {
  id: UUID;
  captured_at: string;
  source: string;
  note?: string | null;
  entries?: FinanceRateSnapshotEntry[];
}

export interface FinanceTreeCreate {
  name: string;
  primary_currency?: string;
  display_order?: number;
  is_default?: boolean;
}

export interface FinanceTreeUpdate {
  name?: string;
  primary_currency?: string;
  display_order?: number;
  is_default?: boolean;
}

export interface FinanceNodeCreate {
  name: string;
  parent_id?: UUID | null;
  currency_code?: string | null;
  display_order?: number;
}

export interface FinanceNodeUpdate {
  name?: string;
  currency_code?: string | null;
  display_order?: number;
}

export interface FinanceSnapshotEntryCreate {
  node_id: UUID;
  amount: string;
  currency_code?: string | null;
  note?: string | null;
}

export interface FinanceSnapshotCreate {
  title?: string | null;
  snapshot_ts?: string | null;
  period_start?: string | null;
  period_end?: string | null;
  primary_currency?: string | null;
  rate_snapshot_id?: UUID | null;
  note?: string | null;
  entries: FinanceSnapshotEntryCreate[];
}

export interface FinanceSnapshotUpdate {
  title?: string | null;
  snapshot_ts?: string | null;
  period_start?: string | null;
  period_end?: string | null;
  primary_currency?: string | null;
  rate_snapshot_id?: UUID | null;
  note?: string | null;
  entries?: FinanceSnapshotEntryCreate[];
}

export interface FinanceRateSnapshotEntryCreate {
  base_currency: string;
  quote_currency: string;
  rate: string;
  source?: string | null;
  captured_at?: string | null;
  is_derived?: boolean;
  metadata?: Record<string, unknown> | null;
}

export interface FinanceRateSnapshotCreate {
  captured_at?: string | null;
  source?: string | null;
  note?: string | null;
  metadata?: Record<string, unknown> | null;
  entries: FinanceRateSnapshotEntryCreate[];
}

export interface FinanceRateSnapshotUpdate {
  captured_at?: string | null;
  source?: string | null;
  note?: string | null;
  metadata?: Record<string, unknown> | null;
  entries?: FinanceRateSnapshotEntryCreate[];
}

type EmptyFinanceListMeta = Record<string, never>;

export type FinanceTreeListResponse = ListResponse<
  FinanceTree,
  EmptyFinanceListMeta
>;
export type FinanceAssetListResponse = ListResponse<
  FinanceAsset,
  EmptyFinanceListMeta
>;
export type FinanceSnapshotListResponse = ListResponse<
  FinanceSnapshot,
  { tree_id?: UUID | null }
>;
export type FinanceRateSnapshotListResponse = ListResponse<
  FinanceRateSnapshot,
  EmptyFinanceListMeta
>;

export const financeApi = {
  listAssets: (params: { page?: number; size?: number } = {}) =>
    http.get<FinanceAssetListResponse>(ENDPOINTS.FINANCE.ASSETS, {
      page: params.page ?? 1,
      size: params.size ?? 200,
    }),
  createAsset: (payload: {
    code: string;
    name?: string | null;
    decimal_places?: number;
    display_order?: number;
    is_default?: boolean;
  }) => http.post<FinanceAsset>(ENDPOINTS.FINANCE.ASSETS, payload),
  updateAsset: (
    assetId: UUID,
    payload: {
      code?: string;
      name?: string | null;
      decimal_places?: number;
      display_order?: number;
    },
  ) => http.patch<FinanceAsset>(ENDPOINTS.FINANCE.ASSET_BY_ID(assetId), payload),
  deleteAsset: (assetId: UUID) =>
    http.delete<void>(ENDPOINTS.FINANCE.ASSET_BY_ID(assetId)),
  listTrees: (
    params: { page?: number; size?: number } = {},
  ) =>
    http.get<FinanceTreeListResponse>(ENDPOINTS.FINANCE.TREES, {
      page: params.page ?? 1,
      size: params.size ?? 100,
    }),
  createTree: (payload: FinanceTreeCreate) =>
    http.post<FinanceTree>(ENDPOINTS.FINANCE.TREES, payload),
  updateTree: (treeId: UUID, payload: FinanceTreeUpdate) =>
    http.patch<FinanceTree>(ENDPOINTS.FINANCE.TREE_BY_ID(treeId), payload),
  deleteTree: (treeId: UUID) =>
    http.delete<void>(ENDPOINTS.FINANCE.TREE_BY_ID(treeId)),
  listRateSnapshots: (params: { page?: number; size?: number } = {}) =>
    http.get<FinanceRateSnapshotListResponse>(
      ENDPOINTS.FINANCE.RATE_SNAPSHOTS,
      {
        page: params.page ?? 1,
        size: params.size ?? 50,
      },
    ),
  createRateSnapshot: (payload: FinanceRateSnapshotCreate) =>
    http.post<FinanceRateSnapshot>(
      ENDPOINTS.FINANCE.RATE_SNAPSHOTS,
      payload,
    ),
  updateRateSnapshot: (
    rateSnapshotId: UUID,
    payload: FinanceRateSnapshotUpdate,
  ) =>
    http.patch<FinanceRateSnapshot>(
      ENDPOINTS.FINANCE.RATE_SNAPSHOT_BY_ID(rateSnapshotId),
      payload,
    ),
  deleteRateSnapshot: (rateSnapshotId: UUID) =>
    http.delete<void>(ENDPOINTS.FINANCE.RATE_SNAPSHOT_BY_ID(rateSnapshotId)),
  getRateSnapshot: (rateSnapshotId: UUID) =>
    http.get<FinanceRateSnapshot>(
      ENDPOINTS.FINANCE.RATE_SNAPSHOT_BY_ID(rateSnapshotId),
    ),
  ensureDefaultTree: (primaryCurrency = "USD") =>
    http.post<FinanceTree>(
      ENDPOINTS.FINANCE.ENSURE_DEFAULT_TREE,
      undefined,
      {
        primary_currency: primaryCurrency,
      },
    ),
  getTree: (treeId: UUID) =>
    http.get<FinanceTree>(ENDPOINTS.FINANCE.TREE_BY_ID(treeId)),
  createNode: (treeId: UUID, payload: FinanceNodeCreate) =>
    http.post<FinanceTreeNode>(ENDPOINTS.FINANCE.TREE_NODES(treeId), payload),
  updateNode: (nodeId: UUID, payload: FinanceNodeUpdate) =>
    http.patch<FinanceTreeNode>(ENDPOINTS.FINANCE.NODE_BY_ID(nodeId), payload),
  deleteNode: (nodeId: UUID) =>
    http.delete<void>(ENDPOINTS.FINANCE.NODE_BY_ID(nodeId)),
  listSnapshots: (treeId: UUID, page = 1, size = 50) =>
    http.get<FinanceSnapshotListResponse>(
      ENDPOINTS.FINANCE.TREE_SNAPSHOTS(treeId),
      { page, size },
    ),
  listAllSnapshots: (page = 1, size = 200) =>
    http.get<FinanceSnapshotListResponse>(ENDPOINTS.FINANCE.SNAPSHOTS, {
      page,
      size,
    }),
  createSnapshot: (treeId: UUID, payload: FinanceSnapshotCreate) =>
    http.post<FinanceSnapshot>(
      ENDPOINTS.FINANCE.TREE_SNAPSHOTS(treeId),
      payload,
    ),
  updateSnapshot: (snapshotId: UUID, payload: FinanceSnapshotUpdate) =>
    http.patch<FinanceSnapshot>(ENDPOINTS.FINANCE.SNAPSHOT_BY_ID(snapshotId), payload),
  deleteSnapshot: (snapshotId: UUID) =>
    http.delete<void>(ENDPOINTS.FINANCE.SNAPSHOT_BY_ID(snapshotId)),
  getSnapshot: (snapshotId: UUID) =>
    http.get<FinanceSnapshot>(ENDPOINTS.FINANCE.SNAPSHOT_BY_ID(snapshotId)),
};

File: src/lifeos_cli/db/models/finance.py
"""Unified finance tree and snapshot models."""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Any
from uuid import UUID

from sqlalchemy import JSON, Boolean, ForeignKey, Index, Integer, Numeric, String, Text, Uuid, text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from lifeos_cli.db.base import Base, SoftDeleteMixin, TimestampedMixin, UUIDPrimaryKeyMixin
from lifeos_cli.db.types import UTCDateTime


class FinanceTree(UUIDPrimaryKeyMixin, TimestampedMixin, SoftDeleteMixin, Base):
    """Reusable finance tree for balance, cashflow, or custom snapshots."""

    __tablename__ = "finance_trees"
    __table_args__ = (
        Index(
            "uq_finance_trees_purpose_name_active",
            "purpose",
            "name",
            unique=True,
            postgresql_where=text("deleted_at IS NULL"),
            sqlite_where=text("deleted_at IS NULL"),
        ),
        Index("ix_finance_trees_purpose_default", "purpose", "is_default"),
        Index("ix_finance_trees_display_order", "display_order"),
    )

    name: Mapped[str] = mapped_column(String(200), nullable=False)
    purpose: Mapped[str] = mapped_column(String(20), nullable=False, default="custom")
    time_mode: Mapped[str] = mapped_column(String(20), nullable=False, default="instant")
    primary_currency: Mapped[str] = mapped_column(String(16), nullable=False, default="USD")
    display_order: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    is_default: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    metadata_json: Mapped[dict[str, Any] | None] = mapped_column("metadata", JSON, nullable=True)

    nodes = relationship(
        "FinanceTreeNode",
        back_populates="tree",
        cascade="all, delete-orphan",
        order_by="FinanceTreeNode.path",
    )
    snapshots = relationship(
        "FinanceSnapshot",
        back_populates="tree",
        cascade="all, delete-orphan",
        order_by="FinanceSnapshot.snapshot_ts.desc()",
    )

    def __repr__(self) -> str:
        return f"FinanceTree(id={self.id!s}, purpose={self.purpose!r}, name={self.name!r})"


class FinanceAsset(UUIDPrimaryKeyMixin, TimestampedMixin, SoftDeleteMixin, Base):
    """Asset code usable by finance trees, snapshots, and exchange rates."""

    __tablename__ = "finance_assets"
    __table_args__ = (
        Index(
            "uq_finance_assets_code_active",
            "code",
            unique=True,
            postgresql_where=text("deleted_at IS NULL"),
            sqlite_where=text("deleted_at IS NULL"),
        ),
        Index("ix_finance_assets_display_order", "display_order", "code"),
    )

    code: Mapped[str] = mapped_column(String(16), nullable=False)
    name: Mapped[str | None] = mapped_column(String(200), nullable=True)
    decimal_places: Mapped[int] = mapped_column(Integer, nullable=False, default=2)
    display_order: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    is_default: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    metadata_json: Mapped[dict[str, Any] | None] = mapped_column("metadata", JSON, nullable=True)

    def __repr__(self) -> str:
        return f"FinanceAsset(id={self.id!s}, code={self.code!r})"


class FinanceTreeNode(UUIDPrimaryKeyMixin, TimestampedMixin, SoftDeleteMixin, Base):
    """Hierarchical finance node shared by balance and cashflow presets."""

    __tablename__ = "finance_tree_nodes"
    __table_args__ = (
        Index("ix_finance_tree_nodes_tree_path", "tree_id", "path", unique=True),
        Index("ix_finance_tree_nodes_parent", "parent_id"),
        Index("ix_finance_tree_nodes_tree_order", "tree_id", "display_order", "created_at"),
    )

    tree_id: Mapped[UUID] = mapped_column(
        Uuid,
        ForeignKey("finance_trees.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    parent_id: Mapped[UUID | None] = mapped_column(
        Uuid,
        ForeignKey("finance_tree_nodes.id", ondelete="CASCADE"),
        nullable=True,
        index=True,
    )
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    currency_code: Mapped[str | None] = mapped_column(String(16), nullable=True)
    path: Mapped[str] = mapped_column(String(1200), nullable=False, default="")
    depth: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    display_order: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    children_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    metadata_json: Mapped[dict[str, Any] | None] = mapped_column("metadata", JSON, nullable=True)

    tree = relationship("FinanceTree", back_populates="nodes")
    parent = relationship(
        "FinanceTreeNode",
        remote_side="FinanceTreeNode.id",
        back_populates="children",
    )
    children = relationship(
        "FinanceTreeNode",
        back_populates="parent",
        cascade="all, delete-orphan",
        order_by="FinanceTreeNode.display_order",
    )
    snapshot_entries = relationship(
        "FinanceSnapshotEntry",
        back_populates="node",
        cascade="all, delete-orphan",
    )

    def __repr__(self) -> str:
        return f"FinanceTreeNode(id={self.id!s}, tree_id={self.tree_id!s}, name={self.name!r})"


class FinanceSnapshot(UUIDPrimaryKeyMixin, TimestampedMixin, SoftDeleteMixin, Base):
    """Snapshot for one finance tree at an instant or over a period."""

    __tablename__ = "finance_snapshots"
    __table_args__ = (
        Index("ix_finance_snapshots_tree_ts", "tree_id", "snapshot_ts"),
        Index("ix_finance_snapshots_tree_period", "tree_id", "period_start", "period_end"),
        Index("ix_finance_snapshots_rate_snapshot", "rate_snapshot_id"),
    )

    tree_id: Mapped[UUID] = mapped_column(
        Uuid,
        ForeignKey("finance_trees.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    rate_snapshot_id: Mapped[UUID | None] = mapped_column(
        Uuid,
        ForeignKey("finance_rate_snapshots.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    title: Mapped[str | None] = mapped_column(String(200), nullable=True)
    snapshot_ts: Mapped[datetime | None] = mapped_column(UTCDateTime(), nullable=True)
    period_start: Mapped[datetime | None] = mapped_column(UTCDateTime(), nullable=True)
    period_end: Mapped[datetime | None] = mapped_column(UTCDateTime(), nullable=True)
    primary_currency: Mapped[str] = mapped_column(String(16), nullable=False)
    rate_snapshot_policy: Mapped[str] = mapped_column(String(32), nullable=False, default="none")
    total_positive: Mapped[Decimal] = mapped_column(Numeric(20, 8), nullable=False, default=0)
    total_negative: Mapped[Decimal] = mapped_column(Numeric(20, 8), nullable=False, default=0)
    net_amount: Mapped[Decimal] = mapped_column(Numeric(20, 8), nullable=False, default=0)
    exchange_rates: Mapped[dict[str, Any] | None] = mapped_column(JSON, nullable=True)
    summary: Mapped[dict[str, Any] | None] = mapped_column(JSON, nullable=True)
    note: Mapped[str | None] = mapped_column(Text, nullable=True)

    tree = relationship("FinanceTree", back_populates="snapshots")
    rate_snapshot = relationship("FinanceRateSnapshot", back_populates="snapshots")
    entries = relationship(
        "FinanceSnapshotEntry",
        back_populates="snapshot",
        cascade="all, delete-orphan",
        order_by="FinanceSnapshotEntry.created_at",
    )

    def __repr__(self) -> str:
        return f"FinanceSnapshot(id={self.id!s}, tree_id={self.tree_id!s})"


class FinanceSnapshotEntry(UUIDPrimaryKeyMixin, TimestampedMixin, SoftDeleteMixin, Base):
    """One node amount captured inside a finance snapshot."""

    __tablename__ = "finance_snapshot_entries"
    __table_args__ = (
        Index(
            "uq_finance_snapshot_entries_snapshot_node_active",
            "snapshot_id",
            "node_id",
            unique=True,
            postgresql_where=text("deleted_at IS NULL"),
            sqlite_where=text("deleted_at IS NULL"),
        ),
        Index("ix_finance_snapshot_entries_node", "node_id"),
    )

    snapshot_id: Mapped[UUID] = mapped_column(
        Uuid,
        ForeignKey("finance_snapshots.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    node_id: Mapped[UUID] = mapped_column(
        Uuid,
        ForeignKey("finance_tree_nodes.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    amount: Mapped[Decimal] = mapped_column(Numeric(20, 8), nullable=False)
    currency_code: Mapped[str] = mapped_column(String(16), nullable=False)
    amount_converted: Mapped[Decimal] = mapped_column(Numeric(20, 8), nullable=False)
    note: Mapped[str | None] = mapped_column(Text, nullable=True)
    is_auto_generated: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)

    snapshot = relationship("FinanceSnapshot", back_populates="entries")
    node = relationship("FinanceTreeNode", back_populates="snapshot_entries")

    def __repr__(self) -> str:
        return f"FinanceSnapshotEntry(id={self.id!s}, node_id={self.node_id!s})"


class FinanceRateSnapshot(UUIDPrimaryKeyMixin, TimestampedMixin, SoftDeleteMixin, Base):
    """Exchange-rate set captured at one time."""

    __tablename__ = "finance_rate_snapshots"
    __table_args__ = (Index("ix_finance_rate_snapshots_captured", "captured_at"),)

    captured_at: Mapped[datetime] = mapped_column(UTCDateTime(), nullable=False)
    source: Mapped[str] = mapped_column(String(64), nullable=False, default="manual")
    note: Mapped[str | None] = mapped_column(Text, nullable=True)
    metadata_json: Mapped[dict[str, Any] | None] = mapped_column("metadata", JSON, nullable=True)

    entries = relationship(
        "FinanceRateSnapshotEntry",
        back_populates="rate_snapshot",
        cascade="all, delete-orphan",
        order_by="FinanceRateSnapshotEntry.base_currency",
    )
    snapshots = relationship("FinanceSnapshot", back_populates="rate_snapshot")

    def __repr__(self) -> str:
        return f"FinanceRateSnapshot(id={self.id!s}, captured_at={self.captured_at!r})"


class FinanceRateSnapshotEntry(UUIDPrimaryKeyMixin, TimestampedMixin, SoftDeleteMixin, Base):
    """One exchange-rate pair inside a rate snapshot."""

    __tablename__ = "finance_rate_snapshot_entries"
    __table_args__ = (
        Index(
            "uq_finance_rate_snapshot_entries_pair_active",
            "rate_snapshot_id",
            "base_currency",
            "quote_currency",
            unique=True,
            postgresql_where=text("deleted_at IS NULL"),
            sqlite_where=text("deleted_at IS NULL"),
        ),
        Index("ix_finance_rate_snapshot_entries_base", "base_currency"),
    )

    rate_snapshot_id: Mapped[UUID] = mapped_column(
        Uuid,
        ForeignKey("finance_rate_snapshots.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    base_currency: Mapped[str] = mapped_column(String(16), nullable=False)
    quote_currency: Mapped[str] = mapped_column(String(16), nullable=False)
    rate: Mapped[Decimal] = mapped_column(Numeric(28, 12), nullable=False)
    source: Mapped[str | None] = mapped_column(String(64), nullable=True)
    captured_at: Mapped[datetime | None] = mapped_column(UTCDateTime(), nullable=True)
    is_derived: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    metadata_json: Mapped[dict[str, Any] | None] = mapped_column("metadata", JSON, nullable=True)

    rate_snapshot = relationship("FinanceRateSnapshot", back_populates="entries")

    def __repr__(self) -> str:
        return (
            "FinanceRateSnapshotEntry("
            f"id={self.id!s}, pair={self.base_currency}/{self.quote_currency})"
        )


__all__ = [
    "FinanceAsset",
    "FinanceRateSnapshot",
    "FinanceRateSnapshotEntry",
    "FinanceSnapshot",
    "FinanceSnapshotEntry",
    "FinanceTree",
    "FinanceTreeNode",
]
